File: Abuse.ch/src/R53AbuseAutoUpdate.js
const AWS = require("aws-sdk");
const response = require("cfn-response");
const https = require("https");

const s3 = new AWS.S3();
const route53resolver = new AWS.Route53Resolver();

const hostfileUrl = "https://urlhaus.abuse.ch/downloads/hostfile/";

async function importList (s3Obj){
    let params = {
      DomainFileUrl: s3Obj,
      FirewallDomainListId: "<REPLACE-WITH-YOUR-DOMAIN-LIST-ID>",
      Operation: "REPLACE"
    };
    
    let res = await route53resolver.importFirewallDomains(params, function(err, data) {
      if (err) console.log(err, err.stack); // an error occurred
      else     console.log("Updating the domain list using " + s3Obj);
    }).promise();
    return res;
}

async function writeToS3(domains){
  let params = {
    Bucket: "<REPLACE-WITH-YOUR-BUCKET-NAME>",
    Key: "autoupdating-dnsfirewall-domains.txt",
    Body: domains.join('\n'),
    ContentType: "text/plain"
  };
  let res = await s3.putObject(params, function(err, data){
                    if (err) console.log(err, err.stack); // an error occurred
                  }).promise();
  if (res) {
    return "https://" + params.Bucket + ".s3.amazonaws.com/" + params.Key;
  }
}

async function deleteFromS3(){
  let params = {
    Bucket: "<REPLACE-WITH-YOUR-BUCKET-NAME>",
    Key: "autoupdating-dnsfirewall-domains.txt"
  };
  let res = await s3.deleteObject(params, function(err, data){
                    if (err) console.log(err, err.stack); // an error occurred
                  }).promise();
  if (res) {
    console.log("Deleted the import file");
  }
}

async function getDomains (){
  let listOfDomains = [];
    console.log("Fetching the list of domains from " + hostfileUrl);
    return new Promise((resolve, reject) => {
      const url = hostfileUrl;
      let dataString = '';
      let post_req = https.request(url, (res) => {
        res.setEncoding("utf8");
        res.on('data', chunk => {
          dataString += chunk;
        });
        res.on('end', () => {
          //console.log(dataString);
          listOfDomains = dataString
            .split(/\r?\n/)
            .filter((line) => line.match(/^\d+/))
            .map((line)=> {return line.replace(/127.0.0.1\t/,'')})
            .filter((line) => line.match(/^(?!:\/\/)(?=.{1,255}$)((.{1,63}\.){1,127}(?![0-9]*$)[a-z0-9-]+\.?)$/));
          console.log("Fetched " + listOfDomains.length + " Domains");
          resolve(listOfDomains);
        });
        res.on('error', (err) => {
          reject(err);
        });
      });
      post_req.write("test");
      post_req.end();
    });
}

exports.handler = async (event, context) => {
  if (event.RequestType == "Delete") {
    await deleteFromS3();
    await response.send(event, context, "SUCCESS");
    return;
  }

  let domains = await getDomains();
  if (domains.length) {
    let s3Obj = await writeToS3(domains, event.key);
    if (s3Obj) {
      let res = await importList(s3Obj);
      if (res) {
        if (event.ResponseURL) await response.send(event, context, response.SUCCESS);
        console.log("Done");
      } else {
        console.log("ERROR - Import failed");
        if (event.ResponseURL) await response.send(event, context, response.FAILED);
      }                
    } else {
      console.log("ERROR - Unable to write domain file to S3");
      if (event.ResponseURL) await response.send(event, context, response.FAILED);
    }
  } else {
    console.log("ERROR - Unable to fetch domain list");
    if (event.ResponseURL) await response.send(event, context, response.FAILED);
  }
  
  return;
};

File: Abuse.ch/src/R53AbuseAutoUpdate.py
import boto3
import json
import urllib.request
import re
from botocore.exceptions import ClientError

s3 = boto3.client('s3')
route53resolver = boto3.client('route53resolver')

HOSTFILE_URL = "https://urlhaus.abuse.ch/downloads/hostfile/"
BUCKET_NAME = "<REPLACE-WITH-YOUR-BUCKET-NAME>"
DOMAIN_LIST_ID = "<REPLACE-WITH-YOUR-DOMAIN-LIST-ID>"
FILE_KEY = "autoupdating-dnsfirewall-domains.txt"

# Compile the pattern once
DOMAIN_PATTERN = re.compile(
    r'^'                           # Start of string
    r'(?!-)'                       # Cannot start with hyphen
    r'(?!.*--)'                    # No consecutive hyphens
    r'(?=.{1,253}$)'              # Total length 1-253 (RFC compliant)
    r'(?:'                         # Non-capturing group for labels
        r'[a-zA-Z0-9]'             # Label must start with alphanumeric
        r'(?:[a-zA-Z0-9-]{0,61}'   # Middle can have hyphens (max 63 total)
        r'[a-zA-Z0-9])?'           # Must end with alphanumeric if >1 char
        r'\.'                      # Followed by dot
    r')+'                          # One or more labels
    r'[a-zA-Z]'                    # TLD starts with letter
    r'(?:[a-zA-Z0-9-]{0,61}'       # TLD middle part
    r'[a-zA-Z0-9])?'               # TLD ends with alphanumeric if >1 char
    r'$'                           # End of string
)

def import_list(s3_obj):
    try:
        res = route53resolver.import_firewall_domains(
            DomainFileUrl=s3_obj,
            FirewallDomainListId=DOMAIN_LIST_ID,
            Operation="REPLACE"
        )
        print(f"Updating the domain list using {s3_obj}")
        return res
    except ClientError as e:
        print(f"Error importing firewall domains: {e}")
        return None

def write_to_s3(domains):
    try:
        s3.put_object(
            Bucket=BUCKET_NAME,
            Key=FILE_KEY,
            Body='\n'.join(domains),
            ContentType="text/plain"
        )
        return f"s3://{BUCKET_NAME}/{FILE_KEY}"
    except ClientError as e:
        print(f"Error writing to S3: {e}")
        return None

def delete_from_s3():
    try:
        s3.delete_object(Bucket=BUCKET_NAME, Key=FILE_KEY)
        print("Deleted the import file")
    except ClientError as e:
        print(f"Error deleting from S3: {e}")

def parse_hostfile(content):
    valid_domains = {
        parts[1].strip()                                    # Extract and clean the domain name
        for line in content.splitlines()                    # Split content into lines
        if not line.startswith('#') and line.strip()        # Skip comments and empty lines
        if (parts := line.split('\t')) and len(parts) >= 2  # Split by tab and ensure we have at least 2 parts
        if DOMAIN_PATTERN.match(parts[1].strip())           # Validate domain with regex pattern
    }
    
    return valid_domains

def get_domains():
    print(f"Fetching the list of domains from {HOSTFILE_URL}")
    try:
        with urllib.request.urlopen(HOSTFILE_URL) as response:
            hostfile_content = response.read().decode('utf-8')
        
        valid_domains = parse_hostfile(hostfile_content)
        print(f"Fetched {len(valid_domains)} domains")
        return list(valid_domains)  # Convert back to list for consistency
    except urllib.error.URLError as e:
        print(f"Error fetching domains: {e}")
        return []
    
def lambda_handler(event, context):
    if event.get('RequestType') == "Delete":
        delete_from_s3()
        return {'statusCode': 200, 'body': json.dumps('SUCCESS')}

    domains = get_domains()
    if not domains:
        print("ERROR - Unable to fetch domain list")
        return {'statusCode': 500, 'body': json.dumps('FAILED - Unable to fetch domain list')}

    s3_obj = write_to_s3(domains)
    if not s3_obj:
        print("ERROR - Unable to write domain file to S3")
        return {'statusCode': 500, 'body': json.dumps('FAILED - Unable to write domain file to S3')}

    res = import_list(s3_obj)
    if res:
        print("Done")
        return {'statusCode': 200, 'body': json.dumps('SUCCESS')}
    else:
        print("ERROR - Import failed")
        return {'statusCode': 500, 'body': json.dumps('FAILED - Import failed')}
